doni/driver/worker/blazar.py

```diff
@@ -1,20 +1,413 @@
+        result = {}
+        try:
+            lease = _call_blazar(
+                context,
+                f"/leases/{lease.get('name')}",
+                method="delete",
+            )
+        except BlazarAPIError as exc:
+            if exc.code == 404:
+                # TODO Host id in lease doesn't exist, what do do?
+                return WorkerResult.Defer(result)
+        else:
+            return WorkerResult.Success(result)
+
     def process(
         self,
         context: "RequestContext",
         hardware: "Hardware",
         availability_windows: "list[AvailabilityWindow]" = None,
-        state_details: "dict" = None,
+        state_details: "dict" = {},
     ) -> "WorkerResult.Base":
-        pass
+        """Main loop for Blazar sync worker.
+
+        This method ensures that an up-to-date blazar host object exists for
+        each physical host in Doni's DB.
+
+        "name" must match the name used by nova to identify the node. In our case
+        it is the hardware uuid, as that is what ironic is passing to nova.
+        blazar uses nova.get_servers_per_host to check if there is an existing
+        server with that name.
+        """
+        # If we know the host_id, then update that host. Otherwise, attempt to create it.
+        host_id = state_details.get("blazar_host_id")
+        if host_id:
+            # TODO: We always update the host. We should add a precondition of some kind.
+            host_result = self._blazar_host_update(context, hardware, host_id)
+        else:
+            # Without a cached host_id, try to create a host. If the host exists,
+            # blazar will match the uuid, and the request will fail.
+            host_result = self._blazar_host_create(context, hardware)
+
+        if isinstance(host_result, WorkerResult.Defer):
+            return host_result  # Return early on defer case
+
+        # Get all leases from blazar
+        leases_to_check = self._blazar_lease_list(context)
+
+        lease_results = []
+        # Loop over all availability windows that Doni has for this hw item
+        for aw in availability_windows or []:
+            new_lease = _blazar_lease_requst_body(aw)
+            # Check to see if lease name already exists in blazar
+            matching_index, matching_lease = _search_leases_for_lease_id(
+                leases_to_check, new_lease
+            )
+            if matching_lease:
+                # Pop each existing lease from the list. Any remaining at the end will be removed.
+                leases_to_check.pop(matching_index)
+                # Update case.
+                if not new_lease.items() <= matching_lease.items():
+                    # If new lease is a subset of old_lease, we don't need to update
+                    lease_results.append(self._blazar_lease_update(context, new_lease))
+            else:
+                lease_results.append(self._blazar_lease_create(context, new_lease))
```

Comment:
Be very careful with stuff like this--the `pop` will mutate the list! Meaning, as soon as it happens once, the indices will be misaligned for the next `pop`. E.g., if you pop index 2, then now what was at index 3 is at index 2, so later if you pop 3, you get a runtime error.

I steer clear of doing anything based on indices for this reason, there are often more robust ways. Here is a suggestion. To find matches, one thing I often do is create a map lookup, mapping from something I know I will have at hand, to the full entity. In your example, we have the lease name. I find it's also useful to break things up into 3 operations: add, update, delete. In this case it would look something like this:

```suggestion
        # Note this is a dict now, mapping names to existing leases in blazar
        existing_leases = {l["name"]: l for l in self._blazar_lease_list(context)}
        desired_leases = [_blazar_lease_requst_body(aw) for aw in availability_windows or []]
        # Transform into matching structure
        desired_leases = {l["name"]: l for l in desired_leases}
        
        # Sets have useful properties/operators like &
        desired_set = set(desired_leases.keys())
        existing_set = set(existing_leases.keys())
        
        lease_results = []
        for l_name in desired_set - existing_set:
            lease_results.append(self._blazar_lease_create(context, desired_leases[l_name]))
        for l_name in desired_set & existing_set:
            if desired_leases[l_name].items() > existing_leases[l_name].items():
                lease_results.append(self._blazar_lease_update(context, desired_leases[l_name]))
        for l_name in desired_set - existing_set:
            lease_results.append(self._blazar_lease_delete(context, existing_leases[l_name]))
```

Reply:
see what I mean:

```
>>> l = [{"foo": "bar"}, {"bar": "baz"}, {"foobar": True}]
>>> l.pop(1)
{'bar': 'baz'}
>>> l
[{'foo': 'bar'}, {'foobar': True}]
>>> l.pop(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: pop index out of range
```